book/layout.md

```diff
@@ -5,529 +5,632 @@ prev: html
+paragraph or header or something like that should take up all the
+horizontal space it can. So it should use the parent's width:
 
 ``` {.python}
-def layout(self, y):
-    self.y = y
-    # ...
+self.w = self.parent.w
 ```
 
-That also lets us remove the hard-coded assignment to `self.y` in
-`__init__`, and while we're at it, let's change `self.x` and `self.w`
-to not be hard-coded:
+Then, each of the children need to be laid out. That means each child
+has to be positioned, and its `layout` method must be called:
 
-``` {.python}
-def __init__(self, parent):
-    # ...
-    self.x = self.parent.x
-    self.y = self.parent.y
+``` {.python indent=8}
+y = self.y
+for child in self.children:
+    child.x = self.x
+    child.y = y
+    child.layout()
+    y += child.h
 ```
 
-Since we're now looking for data in the parent, let's add a new type
-of node to be the root node:
+Note that the call to the child's `layout` method not only asks it to
+compute its size, but also to build more of the layout tree.
+
+The height of an element must include all its children. Since the
+children's heights have already all been summed into `y`, we can just
+subtract its starting value to compute the height:
 
 ``` {.python}
-class Page:
-    def __init__(self):
-        self.x = 13
-        self.y = 13
-        self.w = 774
-        self.children = []
+self.h = y - self.y
 ```
 
-Third and fourth, we need to call `InlineLayout` to handle
-`TextNode`s, plus the `ElementNode`s that are supposed to be inline,
-like `<b>` and `<i>`. The idea is that instead using `BlockLayout` to
-lay out each child node, we\'ll look to see what that node contains.
-If it contains a `TextNode`, or if contains a `<b>` or `<i>` element,
-it will be laid out with `InlineLayout` instead:
+That settles `BlockLayout`; let's now work on `InlineLayout`. Its `x`
+and `y` will be set by its parent, so its `layout` just needs to set
+`w` and `h`:
 
 ``` {.python}
-def is_inline(node):
-    if isinstance(node, TextNode):
-        return true
-    else:
-        return node.tag in ["b", "i"]
-
-def layout(self, y):
-    self.y = y
-    if any(is_inline(child) for child in node.children):
-        layout = InlineLayout(self)
-        self.children.append(layout)
-        layout.layout(node)
-        y += layout.height()
-    else:
+class InlineLayout:
+    def layout(self):
+        self.w = self.parent.w
         # ...
+        self.h = self.cy - self.y
 ```
 
-Let\'s try it out. We\'ll need to tweak `show`:
+Finally even `DocumentLayout` needs some layout code, though since the
+document always starts in the same place it's pretty simple:
 
 ``` {.python}
-page = Page()
-layout = BlockLayout(page)
-layout.layout(nodes)
-maxh = layout.h
-display_list = layout.display_list()
+class DocumentLayout:
+    def layout(self):
+        self.w = WIDTH
+        # ...
+        child.x = self.x = 0
+        child.y = self.y = 0
+        child.layout()
+        self.h = child.h
```

Comment:
Or the height of the viewport instead. This makes it easier to explain how backgrounds of the root elements work, as well as clipping and scrolling of the "DocumentLayout".